src/protocol.c

```diff
@@ -673,4 +710,459 @@ int trilogy_build_ssl_request_packet(trilogy_builder_t *builder, TRILOGY_CAPABIL
     return rc;
 }
 
+int trilogy_build_stmt_prepare_packet(trilogy_builder_t *builder, const char *sql, size_t sql_len)
+{
+    int rc = TRILOGY_OK;
+
+    CHECKED(trilogy_builder_write_uint8(builder, TRILOGY_CMD_STMT_PREPARE));
+
+    CHECKED(trilogy_builder_write_buffer(builder, sql, sql_len));
+
+    trilogy_builder_finalize(builder);
+
+    return TRILOGY_OK;
+
+fail:
+    return rc;
+}
+
+int trilogy_build_stmt_execute_packet(trilogy_builder_t *builder, uint32_t stmt_id, uint8_t flags,
+                                      trilogy_binary_value_t *binds, uint16_t num_binds)
+{
+    int rc = TRILOGY_OK;
+
+    CHECKED(trilogy_builder_write_uint8(builder, TRILOGY_CMD_STMT_EXECUTE));
+
+    CHECKED(trilogy_builder_write_uint32(builder, stmt_id));
+
+    CHECKED(trilogy_builder_write_uint8(builder, flags));
+
+    // apparently, iteration-count is always 1
+    CHECKED(trilogy_builder_write_uint32(builder, 1));
+
+    int i;
+
+    if (num_binds > 0) {
+        if (binds == NULL) {
+            return TRILOGY_PROTOCOL_VIOLATION;
+        }
+
+        uint8_t current_bits = 0;
+
+        for (i = 0; i < num_binds; i++) {
+            if (binds[i].is_null) {
+                current_bits |= 1 << (i % 8);
+            }
+
+            // If we hit a byte boundary, write the bits we have so far and continue
+            if ((i % 8) == 7) {
+                CHECKED(trilogy_builder_write_uint8(builder, current_bits))
+
+                current_bits = 0;
+            }
+        }
+
+        // If there would have been any remainder bits, finally write those as well
+        if (num_binds % 8) {
+            CHECKED(trilogy_builder_write_uint8(builder, current_bits))
+        }
+
+        // new params bound flag
+        CHECKED(trilogy_builder_write_uint8(builder, 0x1));
+
+        for (i = 0; i < num_binds; i++) {
+            CHECKED(trilogy_builder_write_uint8(builder, binds[i].type));
+            CHECKED(trilogy_builder_write_uint8(builder, 0x0));
```

Comment:
I see something about this byte being used to indicate whether the type is unsigned. Is that something we care about here?

Reply:
Totally missed coming back to this, good catch!